data/scenarios/Challenges/hanoi-invariant.sw

```diff
@@ -0,0 +1,47 @@
+def repeat = \c. c; repeat c end;
+def isUnlocked = \e. e == "one" || e == "two" || e == "three" end;
+def unlock = \e.
+  if (e == "blocked one")   {"one"}   {
+  if (e == "blocked two")   {"two"}   {
+  if (e == "blocked three") {"three"} {
+  fail $ "Can not unlock: " ++ e
+  }}}
+end;
+
+repeat (
+me <- scan down;
+case me (\_. return ()) (\e.
+// if
+//  0. I stand on unlocked X
+//  1. place north of me is NOT empty
+// then
+//  - lock X
+if (isUnlocked e)
+{
+    mn <- scan north;
+    case mn (\_. return ()) (\_. grab; place ("blocked " ++ e))
+}
+// if
+//  0. I stand on locked X
+//  1. place north of me is empty
+//  2. all disks are placed
+//  3. other columns are sorted (check "OK")
+// then
+//  - unlock X
+{
+    mn <- scan north;
+    case mn (\_.
+      // TODO: check all disks are placed
+      allSorted <- as self {
+        teleport self (undefined);
+        o1 <- ishere "OK";
+        teleport self (undefined);
+        o2 <- ishere "OK";
+        teleport self (undefined);
+        o3 <- ishere "OK";
+        return (o1 && o2 && o3)
+      };
+      if allSorted {grab; place (unlock e)} {}
```

Comment:
> `grab; place (unlock e)`

@byorgey can this be done atomically? I would like to `swap` the entities without there ever being an empty space.

Reply:
No, because `grab` and `place` are both tangible commands, and you're allowed at most one of those inside an `atomic` block.  However, I think I would be fine with putting in an exception for system robots.  Not quite sure how that exception would work in practice though, since checking of `atomic` blocks is done at typechecking time, not runtime...

We could also just add a special `swap : str -> cmd str` which does a `grab` and a `place` atomically?

Reply:
I think `swap` will be better and also faster since it will only be one command. :+1: 

We can add it later, though the Hanoi tower challenge would look nicer with `swap`.

Reply:
https://github.com/swarm-game/swarm/issues/680

Reply:
> I think swap will be better and also faster...

... and necessary. The last rule makes race conditions much more likely. :weary: